src/test/java/org/opentripplanner/model/TimetableTest.java

```diff
@@ -591,6 +591,61 @@ public void testUpdateWithRequiredDelayPropagationWhenItsRequired() {
     });
   }
 
+  @Test
+  public void testUpdateWithCancellation() {
+    var tripDescriptorBuilder = tripDescriptorBuilder(TRIP_ID);
+    TripUpdate.Builder tripUpdateBuilder = TripUpdate.newBuilder();
+    tripUpdateBuilder.setTrip(tripDescriptorBuilder);
+
+    // Stop 1
+    StopTimeUpdate.Builder stopTimeUpdateBuilder = tripUpdateBuilder.addStopTimeUpdateBuilder(0);
+    stopTimeUpdateBuilder.setStopSequence(1);
+    stopTimeUpdateBuilder.setScheduleRelationship(StopTimeUpdate.ScheduleRelationship.SCHEDULED);
+    StopTimeEvent.Builder stopTimeEventBuilder = stopTimeUpdateBuilder.getArrivalBuilder();
+    stopTimeEventBuilder.setDelay(0);
+    stopTimeEventBuilder = stopTimeUpdateBuilder.getDepartureBuilder();
+    stopTimeEventBuilder.setDelay(0);
+
+    // Stop 2
+    stopTimeUpdateBuilder = tripUpdateBuilder.addStopTimeUpdateBuilder(1);
+    stopTimeUpdateBuilder.setStopSequence(2);
+    stopTimeUpdateBuilder.setScheduleRelationship(StopTimeUpdate.ScheduleRelationship.SKIPPED);
+
+    // Stop 3
+    stopTimeUpdateBuilder = tripUpdateBuilder.addStopTimeUpdateBuilder(2);
+    stopTimeUpdateBuilder.setStopSequence(3);
+    stopTimeUpdateBuilder.setScheduleRelationship(StopTimeUpdate.ScheduleRelationship.SCHEDULED);
+    stopTimeEventBuilder = stopTimeUpdateBuilder.getArrivalBuilder();
+    stopTimeEventBuilder.setDelay(-800);
+    stopTimeEventBuilder = stopTimeUpdateBuilder.getDepartureBuilder();
+    stopTimeEventBuilder.setDelay(-800);
+
+    TripUpdate tripUpdate = tripUpdateBuilder.build();
```

Comment:
I think you can reduce the duplication in this code a lot by using the `TripUpdateBuilder` like this: https://github.com/opentripplanner/OpenTripPlanner/blob/b5d1a9cf20b4e716008185c720f1699414b01883/src/test/java/org/opentripplanner/updater/trip/TimetableSnapshotSourceTest.java#L507-L519

If there are methods missing in that class you can add them.